### PR TITLE
Centralize task creation

### PR DESCRIPTION
**Pull request by DougGregor**
Centralize and clean up task creation facilities for the concurrency runtime. This involves a number of changes:
* Route all task creation through a single, overridable function `swift_task_create_common`. We can back-patch this with the usual runtime "override" mechanism to fix bugs or back-deploy features as necessary.
* Reduce the external entry points for task creation down to three: `swift_task_create` (used for spawning most tasks), `swift_task_create_common` (used when we already have the function pointer we need), and `swift_asyncLet_start` (used to start `async let` tasks). Make the Swift task-launching entry points small, always-emit-into-client shims over `swift_task_create` so we can evolve the API without changing the ABI.
* Use a distinct set of flags for launching tasks (rather than repurposing job flags), and add in support for copying task locals, inheriting other context, enqueuing the task (including on a specified executor), and wiring up async let storage.
* Rework the SIL builtins for task creation to use the new task creation flags and always map down to `swift_task_create`.
* Make `Task` `@frozen` because it's always going to be a wrapper over a single pointer.

Implements rdar://79584250

### CODE REVIEW

File: include/swift/AST/Builtins.def [hunk 1]
```diff
@@ -805,6 +805,17 @@ BUILTIN_MISC_OPERATION(StartAsyncLet, "startAsyncLet", "", Special)
...
+/// createAsyncTask(): (
+///     Int,                            // task-creation flags
+///     Builtin.RawPointer?,            // options (TaskOptionRecord*)
+///     @escaping () async throws -> T  // function
+/// ) -> Builtin.NativeObject
+///
+/// Create a new asynchronous task, given flags, options, and a function to
+/// execute.
+BUILTIN_MISC_OPERATION_WITH_SILGEN(CreateAsyncTask,
+    "createAsyncTask", "", Special)
```
Comment (ktoso) [inline include/swift/AST/Builtins.def:817]: wohoo ❤️

File: stdlib/public/Concurrency/Task.cpp [hunk 2]
```diff
@@ -396,43 +397,70 @@ task_future_wait_resume_adapter(SWIFT_ASYNC_CONTEXT AsyncContext *_context) {
...
+  // Propagate task-creation flags to job flags as appropriate.
+  JobFlags jobFlags(JobKind::Task, taskCreateFlags.getPriority());
+
+  bool isAsyncLetTask = taskCreateFlags.isAsyncLetTask();
+  if (isAsyncLetTask) {
+    jobFlags.task_setIsAsyncLetTask(true);
+    jobFlags.task_setIsChildTask(true);
+  } else {
+    jobFlags.task_setIsChildTask(taskCreateFlags.isChildTask());
+  }
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:424]: ah yeah that's much nicer. I had initially started implementing this by asserting here but makes sense to just infer from asynclet -> must be child. 👍

File: stdlib/public/Concurrency/Task.cpp [hunk 3]
```diff
@@ -396,43 +397,70 @@ task_future_wait_resume_adapter(SWIFT_ASYNC_CONTEXT AsyncContext *_context) {
...
+      executor = cast<ExecutorTaskOptionRecord>(option)->getExecutor();
+      break;
+
+    case TaskOptionRecordKind::TaskGroup:
+      group = cast<TaskGroupTaskOptionRecord>(option)->getGroup();
+      assert(group && "Missing group");
+      jobFlags.task_setIsGroupChildTask(true);
+      break;
+    }
+  }
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:446]: 👍

File: stdlib/public/Concurrency/Task.cpp [hunk 4]
```diff
@@ -649,6 +641,49 @@ getAsyncClosureEntryPointAndContextSize(void *function,
...
+/// until we can eliminate the entry points that operate in terms of job
+/// flags.
+static size_t convertJobFlagsToTaskCreateFlags(size_t rawJobFlags) {
+  JobFlags jobFlags(rawJobFlags);
+  TaskCreateFlags taskCreateFlags;
+  taskCreateFlags.setPriority(jobFlags.getPriority());
+  taskCreateFlags.setIsChildTask(jobFlags.task_isChildTask());
+  taskCreateFlags.setIsAsyncLetTask(jobFlags.task_isAsyncLetTask());
+  return taskCreateFlags.getOpaqueValue();
+}
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:685]: add FIXME? though perhaps we'd be able to remove this still in the same PR?

File: stdlib/public/Concurrency/Task.swift [hunk 5]
```diff
@@ -379,6 +379,97 @@ struct JobFlags {
...
+        bits = bits | 1 << 8
+      } else {
+        bits = (bits & ~(1 << 8))
+      }
+    }
+  }
+
+  /// Whether to copy thread locals from the currently-executing task into the
+  /// newly-created task.
+  var copyThreadLocals: Bool {
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:426]: Hmm... Is this `copyThreadLocalTaskLocals` rather?

File: stdlib/public/Concurrency/Task.swift [hunk 6]
```diff
@@ -379,6 +379,97 @@ struct JobFlags {
...
+        bits = bits | 1 << 11
+      } else {
+        bits = (bits & ~(1 << 11))
+      }
+    }
+  }
+
+  /// Whether to enqueue the newly-created task on the given executor (if
+  /// specified) or the global executor.
+  var enqueueJob: Bool {
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:458]: nice idea :) was quite annoying to deal with the always having to enqueue "on the outside" (e.g. in swift code again)...

**Review by ktoso — APPROVED** [review]
Looks great so far :) Really looking forward to removing all the other entry points 💯

File: stdlib/public/Concurrency/Task.cpp (hunk 4 above)
Comment (DougGregor) [reply stdlib/public/Concurrency/Task.cpp:685]: Yes, it'll go away in a subsequent commit. I should  have marked this PR as a draft.

File: stdlib/public/Concurrency/Task.swift (hunk 5 above)
Comment (DougGregor) [reply stdlib/public/Concurrency/Task.swift:426]: D'oh! Yes, task locals, not thread locals.

File: stdlib/public/Concurrency/Task.swift (hunk 6 above)
Comment (DougGregor) [reply stdlib/public/Concurrency/Task.swift:458]: Yeah, I'm trying to get the Swift entry points down to "set of flags, call into the C++ code".

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
The macOS full test hit the playground logger issue but was otherwise good. Trying the smoke test

**Comment by DougGregor** [comment]
@swift-ci please test

File: lib/IRGen/GenBuiltin.cpp [hunk 7]
```diff
@@ -270,16 +270,26 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
...
     auto flags = args.claimNext();
-    auto taskOptions = args.claimNext();
+    auto taskGroup =
+        (Builtin.ID == BuiltinValueKind::CreateAsyncTaskInGroup)
+        ? args.claimNext()
+        : nullptr;
+    llvm::Value *taskOptions =
+        (Builtin.ID == BuiltinValueKind::CreateAsyncTask)
+        ? args.claimNext()
+        : llvm::ConstantInt::get(IGF.IGM.SwiftTaskOptionRecordPtrTy, 0);
```
Comment (ktoso) [inline lib/IRGen/GenBuiltin.cpp]: TIL: oh! that's good to know

File: include/swift/ABI/TaskOptions.h [hunk 8]
```diff
@@ -80,14 +80,14 @@ class TaskGroupTaskOptionRecord : public TaskOptionRecord {
 /// executor should be used instead, most often this may mean the global
 /// concurrent executor, or the enclosing actor's executor.
 class ExecutorTaskOptionRecord : public TaskOptionRecord {
-  ExecutorRef *Executor;
+  ExecutorRef Executor;
```
Comment (ktoso) [inline include/swift/ABI/TaskOptions.h]: 👍

File: include/swift/ABI/MetadataValues.h [hunk 9]
```diff
@@ -2029,11 +2029,12 @@ class TaskCreateFlags : public FlagSet<size_t> {
...
-    Task_IsAsyncLetTask   = 9,
-    Task_CopyThreadLocals = 10,
-    Task_InheritContext   = 11,
-    Task_EnqueueJob       = 12,
+    Task_IsChildTask                              = 8,
+    Task_IsAsyncLetTask                           = 9,
+    Task_CopyThreadLocals                         = 10,
+    Task_InheritContext                           = 11,
+    Task_EnqueueJob                               = 12,
+    Task_AddPendingGroupTaskUnconditionally       = 13,
```
Comment (ktoso) [inline include/swift/ABI/MetadataValues.h:2037]: ah, excellent to put it as a flag 👍

File: lib/IRGen/GenCall.cpp [hunk 10]
```diff
@@ -3861,16 +3861,14 @@ llvm::Value *irgen::emitTaskCreate(
...
     llvm::Value *taskFunction,
     llvm::Value *localContextInfo,
     SubstitutionMap subs) {
-  taskOptions = IGF.Builder.CreateBitOrPointerCast(
-      taskOptions, IGF.IGM.SwiftTaskOptionRecordPtrTy);
-
   // If there is a task group, emit a task group option structure to contain
   // it.
+  llvm::Value *taskOptions = llvm::ConstantInt::get(
+      IGF.IGM.SwiftTaskOptionRecordPtrTy, 0);
```
Comment (ktoso) [inline lib/IRGen/GenCall.cpp:3871]: I see, that's much easier... thanks :)

File: stdlib/public/Concurrency/Task.swift [hunk 11]
```diff
@@ -36,9 +36,12 @@ import Swift
 /// individually schedulable as jobs.  Jobs are generally not interacted
 /// with by end-users directly, unless implementing a scheduler.
 @available(SwiftStdlib 5.5, *)
+@frozen
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:39]: ⛄ ❤️

**Review by ktoso — APPROVED** [review]
LGTM, thanks a lot for taking this over! Learnt a lot by reading this 👍 

No concerns, all looks great 👍

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Pull request merged**